File: aoc24/3/main.clj
(def data (slurp "in.txt"))

(def mult-pattern #"mul\((\d+),(\d+)\)")
(def do-pattern #"do\(\)")
(def dont-pattern #"don't\(\)")

(defn get-idxs [pattern data]
  (let [matcher (re-matcher pattern data)]
    (loop [idxs []]
      (if (.find matcher)
        (recur (conj idxs (.start matcher)))
        idxs))))

(def do-idxs (get-idxs do-pattern data))
(def dont-idxs (get-idxs dont-pattern data))

(defn mult-instruction [[idx n1 n2]]
  (let [do-idx (last (cons 0 (filter #(< % idx) do-idxs)))
        dont-idx (last (cons 0 (filter #(< % idx) dont-idxs)))]
    (if (>= do-idx dont-idx)
      (* n1 n2)
      0)))

(def ans (->> data
              (re-seq mult-pattern)
              (map #(* (Integer/parseInt (nth % 1)) (Integer/parseInt (nth % 2))))
              (reduce +)))

(def mults (re-matcher mult-pattern data))
(def mult-idxs (loop [idxs []]
                 (if (.find mults)
                   (recur (conj idxs (list (.start mults)
                                           (Integer/parseInt (.group mults 1))
                                           (Integer/parseInt (.group mults 2)))))
                   idxs)))

(def ans2 (->> mult-idxs
               (map mult-instruction)
               (reduce +)))

(println ans)
(println ans2)

File: aoc24/3/main.py
import re
import sys

data = open(sys.argv[1]).read().strip()

ans, ans2 = 0, 0
for line in [data]:
    mults = re.finditer('mul\((\d+),(\d+)\)', line)
    dos = list(re.finditer('do\(\)', line))
    donts = list(re.finditer('don\'t\(\)', line))
    for mult in mults:
        do_idx = ([0]+[do.start() for do in dos if do.start() < mult.start()])[-1]
        dont_idx = ([0]+[dont.start() for dont in donts if dont.start() < mult.start()])[-1]
        if do_idx >= dont_idx:
            ans2 += int(mult.group(1)) * int(mult.group(2))
        ans += int(mult.group(1)) * int(mult.group(2))

print(ans)
print(ans2)
